### PR TITLE
[move-ide] Added support for init function auto-completion

### PR DESCRIPTION
**Pull request by awelc**
## Description 

This PR adds preliminary support for auto-completing init function definition. It contains the beginning of the proposed fine-grain error recovery in the parser, currently only implemented for function definitions. The gist of the idea is that we parse function definition as much as we can, and regardless of whether the parsing was fully successful or not, we still return a correct function `Function` AST node if we can (i.e., if we correctly parse attributes, `fun` keyword, and function name).

In particular, this would generate a correct `Function` AST node, for function `ini` even though its definition is incomplete:
```
module 0x0::some_module {
    fun ini
}
```

What this allows us to do, is context-sensitive auto-completion for `init` functions using LSP concept of snippets. In particular, when triggering auto-completion in the code fragment above right after `ini`, the IDE will recognize that we are trying to auto-complete a function whose name starts with `ini` and will offer a code snippet expanding this to a full init function definition (we will be able to omit `sui::tx_context` prefix once auto-imports are in place):
```
module 0x0::some_module {
    fun init(ctx: &mut sui::tx_context::TxContext) {   
    }
}
```

Furthermore, if the module contains a struct definition being a one-time-witness candidate, the IDE will be smart enough to recognize this and offer an expanded snippet:
```
module 0x0::some_module {
    struct SOME_MODULE has drop {}
    fun init(witness: SOME_MODULE, ctx: &mut sui::tx_context::TxContext) {   
    }
}
```

Finally, this snippet will not be offered for all identifiers with a given prefix, only those that represent function definitions. In particular, it will not do it for local variable declaration in the following code fragment:
```
module 0x0::some_module {
    fun foo() {
        let ini   
    }
}
```

The main current limitation is that at this point auto-completion must be preceded by a save (to rebuild symbols information in the LSP server) and manual re-triggering of the auto-completion command. In order to support a smooth user experience, we need to add support for the compiler to use a virtual file system (files stored in memory).

## Test Plan 

Added a compiler test and a symbolicator test for this feature.

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
Developers might see more compiler diagnostics as selected parsing errors no longer prevent compilation and diagnostics from the compiler reaching later compilation stages where additional diagnostics may be generated.

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs [hunk 1]
```diff
@@ -2327,36 +2327,123 @@ fn parse_function_decl(
...
     } = modifiers;
 
+    // certain portions of of the partially constructed function must be updated on return
+    fn update_partial(start_loc: usize, context: &mut Context, mut result: Function) -> Function {
+        result.loc = make_loc(
+            context.tokens.file_hash(),
+            start_loc,
+            context.tokens.previous_end_loc(),
+        );
+        result.body = sp(context.tokens.current_token_loc(), FunctionBody_::Native);
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/parser/syntax.rs:2345]: I'm dubious about marking this native, as it will likely produce very strange results. Inserting `abort 0` or something is safer. We could even consider adding a `FunctionBody_::Error` form.

Comment (awelc) [reply external-crates/move/crates/move-compiler/src/parser/syntax.rs:2345]: This is exactly the kind of feedback I am looking for! I was wondering about it myself (also because we introduce an essentially fake location here), but did not want to go to crazy with the changes without getting feedback first.

In a similar vein (though perhaps more acceptable) is that an incomplete function actually has a `Type_::Unit` return type (rather then perhaps something like `Type_::Error`).

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs [hunk 2]
```diff
@@ -2327,36 +2327,123 @@ fn parse_function_decl(
...
+        result
+    }
+
     // "fun" <FunctionDefName>
     consume_token(context.tokens, Tok::Fun)?;
-    let name = FunctionName(parse_identifier(context)?);
-    let type_parameters = parse_optional_type_parameters(context)?;
+    let nid = parse_identifier(context)?;
+    let name = FunctionName(nid);
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/parser/syntax.rs:2352]: Why the change here?

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs [hunk 3]
```diff
@@ -2327,36 +2327,123 @@ fn parse_function_decl(
...
+        entry,
+        macro_,
+        signature: FunctionSignature {
+            type_parameters: vec![],
+            parameters: vec![],
+            return_type: sp(name.loc(), Type_::Unit),
+        },
+        name,
+        body: sp(context.tokens.current_token_loc(), FunctionBody_::Native),
+    };
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/parser/syntax.rs:2373]: A rather nit-picky nit. But thoughts on making a local builder here? 

Since there aren't that many fields we would need to set, we could do:
```
let mut type_parameters = None;
let mut parameters = None;
let mut return_type = None;
let mut body = None;
let mk_result = |current_loc| Function {
    loc,
    visibility: visibility.unwrap_or(Visibility::Internal),
    entry,
    macro_,
    signature: FunctionSignature {
        type_parameters: type_parameters.unwrap_or_default(),
        parameters: parameters.unwrap_or_default(),
        return_type: return_type.unwrap_or_else(|| sp(name.loc(), Type_::UnresolvedError)),
    },
    name,
    body: body.unwrap_or_else(|| sp(current_loc, FunctionBody_::UnresolvedError)),
  };

Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/parser/syntax.rs:2373]: It's not _that_ much different, but feels a lot more intentional, which makes me happy :)

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs (hunk 1 above)
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/parser/syntax.rs:2345]: +1 For something like `FunctionBody_::Error`
Though you could probably achieve the same thing by doing `FunctionBody_::Defined(...UnresolvedError...)`

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs [hunk 4]
```diff
@@ -2327,36 +2327,123 @@ fn parse_function_decl(
     start_loc: usize,
     modifiers: Modifiers,
     context: &mut Context,
-) -> Result<Function, Box<Diagnostic>> {
+) -> Result<(Function, Option<Box<Diagnostic>>), Box<Diagnostic>> {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/parser/syntax.rs:2330]: This feels strange to me. Can we not just log the diagnostic if we have it? Why bother passing it back instead of just logging it?

Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/parser/syntax.rs:2330]: I guess what I'm suggesting is that the function parsing can locally handle trying to advance the parser to the next valid state, much like we did in the lexer, and hide the error handling complexity from the caller

Comment (awelc) [reply external-crates/move/crates/move-compiler/src/parser/syntax.rs:2365]: I was not sure what the best way was to structure this, but you are obviously right - we don't really have to pass the diagnostics on the success case when we can log them on the spot

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs (hunk 3 above)
Comment (awelc) [reply external-crates/move/crates/move-compiler/src/parser/syntax.rs]: Done! Though I did turn it to a macro as it looks like Rust does not allow to modify variables after they are captured by the closure.

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs (hunk 1 above)
Comment (awelc) [reply external-crates/move/crates/move-compiler/src/parser/syntax.rs]: > +1 For something like `FunctionBody_::Error` Though you could probably achieve the same thing by doing `FunctionBody_::Defined(...UnresolvedError...)`

I went this route (actually just provided empty `Sequence` as the function body) as this does not require changes to subsequent passes (to handle the new type of `UnresolvedError`).

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/global_access_exists_invalid.exp [hunk 5]
```diff
@@ -12,6 +12,14 @@ error[E03011]: invalid use of reserved name
 9 │     fun freeze(): u64 { 0 }
   │         ^^^^^^ Invalid function name 'freeze'. 'freeze' is restricted and cannot be used to name a function
 
+warning[W09002]: unused variable
+   ┌─ tests/move_2024/parser/global_access_exists_invalid.move:11:11
+   │
+11 │     fun t(account: &signer) acquires Self::R {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/parser/global_access_exists_invalid.exp:18]: nit, could you add an underscore here

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/global_access_exists_invalid.exp [hunk 6]
```diff
@@ -21,3 +29,12 @@ error[E01002]: unexpected token
    = Access chains that start with '::' must be one of the following forms: 
        '::<address>::<module>', '::<address>::<module>::<member>'
 
+error[E01002]: unexpected token
+   ┌─ tests/move_2024/parser/global_access_exists_invalid.move:13:32
+   │
+13 │         let _ : bool = ::exists<Self::R>(0x0);
+   │                                ^
+   │                                │
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/parser/global_access_exists_invalid.exp:37]: How did we end up expecting a module member here?

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/global_access_pack_invalid.exp [hunk 7]
```diff
@@ -7,3 +7,12 @@ error[E01002]: unexpected token
...
       '::<address>::<module>', '::<address>::<module>::<member>'
 
+error[E01002]: unexpected token
+  ┌─ tests/move_2024/parser/global_access_pack_invalid.move:3:13
+  │
+3 │         ::S { }
+  │             ^
+  │             │
+  │             Unexpected '{'
+  │             Expected a module member: 'spec', 'use', 'friend', 'const', 'fun', or 'struct'
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/parser/global_access_pack_invalid.exp:17]: Similar question as above

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/global_access_value_invalid.exp [hunk 8]
```diff
@@ -7,3 +7,21 @@ error[E01002]: unexpected token
   = Access chains that start with '::' must be one of the following forms: 
       '::<address>::<module>', '::<address>::<module>::<member>'
 
+error[E04007]: incompatible types
+  ┌─ tests/move_2024/parser/global_access_value_invalid.move:4:5
+  │
+2 │     fun foo(): u64 {
+  │                --- Expected: 'u64'
+3 │         1 + ::global_value
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/parser/global_access_value_invalid.exp:15]: This looks a bit suspicious

**Review by tnowacki — COMMENTED** [review]
I will take a deeper look tomorrow, but most of these parser errors look really suspicious, like something has gone wrong with the thing that is supposed to consume tokens until the start of the next module member

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/global_access_value_invalid.exp (hunk 8 above)
Comment (awelc) [reply external-crates/move/crates/move-compiler/tests/move_2024/parser/global_access_value_invalid.exp:15]: This is because the function definition is parsed correctly as having the `u64` return type but the body was not parsed correctly and (being treated as empty), typing throws an error on incompatible return type. I adjusted all tests (including this one) to only report parsing errors.

**Comment by awelc** [comment]
> I will take a deeper look tomorrow, but most of these parser errors look really suspicious, like something has gone wrong with the thing that is supposed to consume tokens until the start of the next module member

Generally speaking, the weird errors were a consequence of not passing the error back to top level (recent change) to avoid modifying function signatures (return types). In other words, once a partial function was constructed (partial due to parsing error at some point during function parsing), it was returned to as a valid module member and since no info about the error was passes, next member's parsing was immediately attempted (which would result in another immediate error on the next token).

The current version compromises on this design a bit and  keeps "error state" in the parsing context for the parser to skip tokens when error is encountered and yet a member is parsed (partially) correctly. This way at least we don't have to keep this state in the AST nor propagate it via return types.

**Comment by awelc** [comment]
Superseded by https://github.com/MystenLabs/sui/pull/17137 and https://github.com/MystenLabs/sui/pull/16673
